src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/PatternGridPeripheral.java

```diff
@@ -0,0 +1,302 @@
+            patternStack.set(DataComponents.INSTANCE.getCraftingPatternState(), craftingState);
+
+            // Kismet: a bad recipe will be caught automatically
+            Map<String, Object> result;
+            try {
+                result = getDetailsForItem(patternStack);
+            } catch (Exception e) {
+                throw new RuntimeException("Bad recipe");
+            }
+
+            turtle.getInventory().removeItem(turtle.getSelectedSlot(), 1);
+            turtle.getInventory().setItem(destinationSlot, patternStack);
+
+            return MethodResult.of(true, result);
+        } catch (Exception e) {
+            // I like exceptions but the other peripherals in this mod don't use them. So just return errors as strings.
+            return MethodResult.of(false, e.getMessage());
+        }
+    }
+
+    // -----------------------------------------------------------------------------------------------------------------
+
+    // Build a processing pattern from a list of ingredients.
+    @LuaFunction(mainThread = true)
+    public MethodResult buildProcessing(Map<?, ?> recipeInput) {
+        ITurtleAccess turtle = this.getPeripheralOwner().getTurtle();
+
+        try {
+            verifyPatternsInSlot(turtle.getSelectedSlot());
+            int destinationSlot = getFreeSlotAfterBuild(turtle.getSelectedSlot());
+
+            ItemStack patternStack = new ItemStack(Items.INSTANCE.getPattern());
+            PatternState patternState = new PatternState(UUID.randomUUID(), PatternType.PROCESSING);
+            patternStack.set(DataComponents.INSTANCE.getPatternState(), patternState);
+
+            List<Optional<ProcessingPatternState.ProcessingIngredient>> ingredients = new ArrayList<>();
+            List<Optional<ResourceAmount>> results = new ArrayList<>();
+
+            /* Attempt to parse our recipe. We should be passed in a table that looks like this:
+             *
+             * {
+             *     inputs = {
+             *         [1] = { name = "<item>", count = <int> },
+             *         [2] = { name = "<item>", alts = { [1] = "<tag>", ... }, count = <int> },
+             *         ...
+             *     },
+             *     outputs = {
+             *         [1] = { name = "<item>", count = <int> },
+             *         ...
+             *     }
+             * }
+             *
+             * Note that input names can be passed as a string or as an array of strings (for alternates).
+             * That's a lot to validate, so barring better options, we'll just catch and error out.
+             */
+            try {
+                // Lua sends in all numbers as doubles. We don't really care about the index.
+                Map<Double, Map<String, Object>> inputMap = (Map<Double, Map<String, Object>>) recipeInput.get("inputs");
+                for (Map<String, Object> thisInput : inputMap.values()) {
+                    int thisCount = ((Double) thisInput.get("count")).intValue();
```

Comment:
Don't use type convert here. Instead use `LuaTable` wrapper that CC:T provides.

Reply:
> CC won't accept a `LuaTable` as a method argument

You don't, you use `ObjectLuaTable` wrap the map argument to get `LuaTable` instance.

`LuaTable` operations will throw `LuaException` when type is not match, which is relative more clear than Java class cast exception, and you also don't need wrap the entire method in a try catch anymore.